packages/grafana-data/src/transformations/transformers/groupingToMatrix.ts

```diff
@@ -0,0 +1,105 @@
+import { map } from 'rxjs/operators';
+
+import { DataFrame, DataTransformerInfo, Field, FieldType } from '../../types';
+import { DataTransformerID } from './ids';
+import { MutableDataFrame } from '../../dataframe';
+import { getFieldDisplayName } from '../../field/fieldState';
+
+export interface GroupingToMatrixTransformerOptions {
+  columnField?: string;
+  rowField?: string;
+  valueField?: string;
+}
+
+const DEFAULT_COLUMN_FIELD = 'Time';
+const DEFAULT_ROW_FIELD = 'Time';
+const DEFAULT_VALUE_FIELD = 'Value';
+
+export const groupingToMatrixTransformer: DataTransformerInfo<GroupingToMatrixTransformerOptions> = {
+  id: DataTransformerID.groupingToMatrix,
+  name: 'Grouping to Matrix',
+  description: 'Groups series by field and return a matrix visualisation',
+  defaultOptions: {
+    columnField: DEFAULT_COLUMN_FIELD,
+    rowField: DEFAULT_ROW_FIELD,
+    valueField: DEFAULT_VALUE_FIELD,
+  },
+
+  operator: (options) => (source) =>
+    source.pipe(
+      map((data) => {
+        const columnFieldMatch = options.columnField || DEFAULT_COLUMN_FIELD;
+        const rowFieldMatch = options.rowField || DEFAULT_ROW_FIELD;
+        const valueFieldMatch = options.valueField || DEFAULT_VALUE_FIELD;
+
+        // Accept only single queries
+        if (data.length !== 1) {
+          return data;
+        }
+
+        const frame = data[0];
+        const keyColumnField = findKeyField(frame, columnFieldMatch);
+        const keyRowField = findKeyField(frame, rowFieldMatch);
+        const valueField = findKeyField(frame, valueFieldMatch);
+        const rowColumnField = `${rowFieldMatch}\\${columnFieldMatch}`;
+
+        if (!keyColumnField || !keyRowField || !valueField) {
+          return data;
+        }
+
+        const columnValues = [...new Set(keyColumnField.values.toArray())];
+        const rowValues = [...new Set(keyRowField.values.toArray())];
```

Comment:
It might be a good idea to change these to to something like the example below because it will skip one copy of the values array from being created.

```suggestion
        const columnValues = uniqueValues(keyColumnField.values);
        const rowValues = uniqueValues(keyRowField.values);
```
And add a helper function like this:

```ts
function uniqueValues(values: Vector): any[] {
  const unique = new Set();

  for (let index = 0; index < values.length; index++) {
    unique.add(values.get(index));
  }

  return Array.from(unique);
}
```

Reply:
I think we probably could make this transformation even more efficient but this is good enough for now.

Reply:
I've applied the suggested change and strongly agree with the possibility for improvements :)